web/packages/teleport/src/services/api/api.ts

```diff
@@ -111,6 +111,7 @@ const api = {
       ...params,
       headers: {
         ...params.headers,
+        ...defaultRequestOptions.headers,
         [MFA_HEADER]: JSON.stringify({
           webauthnAssertionResponse: webauthnResponse,
         }),
```

Comment:
This looks incorrect. 🧐 We overwrite headers passed to `fetchJson` through `params.headers` with default headers and then we add an MFA header. Shouldn't it be like this:

```suggestion
      headers: {
        ...defaultRequestOptions.headers,
        ...params.headers,
        [MFA_HEADER]: JSON.stringify({
          webauthnAssertionResponse: webauthnResponse,
        }),
```

to match what happens in `fetch`?

Furthermore, it appears that `fetchJson` calls `fetch` which already applies `defaultRequestOptions`. I feel like the defaults should be applied in just a single place, otherwise it makes the process of applying them harder to understand.

This code "looks" arguably more correct:

```typescript
// fetchJson
    const paramsWithMfaHeader = {
      ...params,
      headers: {
        ...params.headers,
        [MFA_HEADER]: JSON.stringify({
          webauthnAssertionResponse: webauthnResponse,
        }),
      },
    };

// fetch
    const options = {
      ...defaultRequestOptions,
      ...customOptions,
      headers: {
        ...defaultRequestOptions.headers,
        ...customOptions['headers'],
        ...getAuthHeaders(),
      },
    };
```

But does it work? It's pretty much the same as the current implementation on master.

If not, then could you investigate why the code on this branch works? I don't think it's generally correct as on this branch, `fetch` does not merge received headers with the default ones, it always overrides headers from `defaultRequestOptions` with `customOptions`. So if you have any function that passes some custom headers, it needs to repeat the default headers as well. I don't understand why correct working of the app would be predicated on not including `Accept: 'application/json'` and `Content-Type: 'application/json; charset=utf-8'`.

Reply:
Sorry perhaps my wording was incorrect.

When passing in a custom header:

This is options with your suggestion:
![image](https://github.com/gravitational/teleport/assets/43280172/3b763308-f030-4873-aaa2-d0398d166916)

This is options with the current (the PR):
<img width="1149" alt="image" src="https://github.com/gravitational/teleport/assets/43280172/f1c6cda4-41cf-446f-877e-e8b8c8a252dc">

the issue was we are spreading the `header` portion which includes the `content-type` from the default headers and there is an api call where I need the browser to [infer](https://github.com/gravitational/teleport/blob/37f3ed437cd1dafe1f6e8fe48183a1e88a946eb7/web/packages/teleport/src/services/api/api.ts#L41) the content-type for form submissions (so i need this field omitted)

Reply:
to me it sounds right that if you are passing in a custom options, you should ensure the other fields are as expected.... but maybe that's more prone to errors?

Reply:
Ah, okay, so it sounds like we need the following behavior:

* If you don't pass any custom options, the request should use `defaultRequestOptions`.
* If you pass any key in custom options, it should completely override the equivalent key of `defaultRequestOptions`.
   * So if you pass custom headers, you're on your own and you cannot depend on headers from `defaultRequestOptions` being merged with what you provided.

Then your implementation appears indeed to be correct. Could you just document the behavior in a JSDoc for both `fetch` and `fetchJson`? It'd be extra nice to add some tests for that too so we don't run into this problem in the future.